frontend/dockerfile: some minor refactor and optimizations

thaJeztah opened the pull request:
See individual commits for details;

### frontend/dockerfile: remove isSSHMountsSupported, isSecretMountsSupported

These became obsolete in https://github.com/moby/buildkit/commit/906e34510de0f4e39a7f79173d110cff10c90f3b, which always enabled them.

### frontend/dockerfile: parseBuildStageName(): pre-compile regex

Before/After:
    BenchmarkParseBuildStageName-10     180793      5972   ns/op    8355 B/op     126 allocs/op
    BenchmarkParseBuildStageName-10    1980735       605.5 ns/op      48 B/op       3 allocs/op

Also renamed a variable that collided with an import.

### frontend/dockerfile/parser: remove redundant concat

### frontend/dockerfile/shell: use strings.Equalfold

Also fix some grammar issues in the GoDoc, and some nits in the tests.

Before/after:

     BenchmarkToUpper-10          11979394            96.31 ns/op          10 B/op           2 allocs/op
     BenchmarkEqualFold-10        40687216            29.65 ns/op           0 B/op           0 allocs/op

### frontend/dockerfile: define types for enums

This is makes it easier to discover what values are acceptable;

- Generated docs shows them nicely together.
- Intent is clearer (these are all the `MountType` options).
- Intent on fields and arguments is clearer (`func(t MountType)`, `struct Foo{Field: MountType}`),
  but still allows for using / passing a regular string (without casting); validation
  is already handled for these, so invalid strings would (should) already be handled.
- With these types, both docs at pkg.go.dev and IDEs allow for easier navigating
  (click the `MountType` in the examples above to navigate to the `MountType`
  definition, and the options (`const`s) below them.

<img width="355" alt="Screenshot 2022-10-19 at 22 06 33" src="https://user-images.githubusercontent.com/1804568/196799667-7203abc2-9917-4672-a064-0efcca306c24.png">

### frontend/dockerfile: provide suggestions for mount share mode

similar to df9781b46c4ed4f4e7146c2241c0cb513167c2c5

### frontend/dockerfile: move check for cache-sharing

The `m.Type == MountTypeSecret` has a more specific check for this option,
which would not be hit because this check was before it.

Move the check until after the validation for secret mounts, so that the
specific error can be returned, instead of the generic one.

### frontend/dockerfile: parseMount() use strings.Cut(), and some minor cleanup

- use strings.Cut() which simplifies handling of key/value pairs
- remove some redundant intermediate variables
- combined "fileInfoAllowed" if-statement for easier reading

### frontend/dockerfile: parseExtraHosts(): use strings.Cut()

### frontend/dockerfile: BFlags.Parse(): use strings.Cut()

Also some minor cleanup in the boolean value parsing; using a switch
instead of if/elseif/else.

aaronlehmann commented on frontend/dockerfile/instructions/commands_runmount.go:
```diff
@@ -11,25 +11,33 @@ import (
-const MountTypeTmpfs = "tmpfs"
-const MountTypeSecret = "secret"
-const MountTypeSSH = "ssh"
+type MountType = string
```
nit: I think we want

```suggestion
type MountType string
```

The `=` defines a type alias, which are meant for refactoring code across packages.

thaJeztah replied:
So, this was mostly intentional. I'm using aliases (see them as "pseudo" "types") for these as throughout the code these are _actually_ just strings. The consts in the code are mostly used as a "utility" for convenience, but no strong type enforcement is currently happening.

I picked aliases to make the change less disruptive. These values are used for parsing strings passed as input, and switching to use a strong type would really complicate those, e.g. the below would require quite a refactor;

```go
func parseVal(val string) {
    switch strings.Tolower(val) {
    case FooEnabled:
        // something
    case BarEnabled:
        // something else
    default:
        panic("unknown option")
    }
}
```

The "pseudo types" _do_ still bring some of the benefits;

- Generated docs shows them nicely together
- Intent is clearer (these are all the `MountType` options)
- Intent on fields and arguments (`setMountType(t MountType)`), but still allowing to pass a regular string (without casting); validation is already handled for these, so invalid strings would (_should_) already be handled.
- Same for fields (`struct Foo{Field: MountType}`)
- IDE assistance (click the `MountType` above to navigate to all `MountType` options); same in docs (click the field navigates to the type definition, and the consts defined)

That said, we _could_ switch to using a strong type at some point (if we see a strong need for that).

jedevc replied:
I'd agree with @aaronlehmann here - I think we should have a strong preference for creating a new type, consistent with other places where we have enums: `UsageRecordType`, `UlimitName`, `CapID`, etc. The only place where we properly use alias types is for the shared `Result` generic across different packages.

I don't think using a new type should increase the commit footprint significantly - though even if it does, I think it's worth the consistency with the rest of the codebase.

aaronlehmann commented on frontend/dockerfile/instructions/commands_runmount.go:
```diff
@@ -11,25 +11,33 @@ import (
-const MountSharingShared = "shared"
-const MountSharingPrivate = "private"
-const MountSharingLocked = "locked"
+type ShareMode = string
```
```suggestion
type ShareMode string
```

aaronlehmann commented on frontend/dockerfile/instructions/commands_runnetwork.go:
```diff
@@ -4,13 +4,15 @@ import (
 	"github.com/pkg/errors"
 )
 
+type NetworkMode = string
```
```suggestion
type NetworkMode string
```

aaronlehmann commented on frontend/dockerfile/instructions/parse.go:
```diff
@@ -362,13 +362,14 @@ func parseFrom(req parseRequest) (*Stage, error) {
-		if ok, _ := regexp.MatchString("^[a-z][a-z0-9-_\\.]*$", stageName); !ok {
-			return "", errors.Errorf("invalid name for build stage: %q, name can't start with a number or contain symbols", args[2])
+		if !validStageName.MatchString(stageName) {
+			return "", errors.Errorf("invalid name for build stage: %q, name can't start with a number or contain symbols", stageName)
```
nit: this will include the lowercased version in the error message - is that what we want?

thaJeztah replied:
Oh! Good catch; I initially made the regex itself case-insensitive (adding `(?i)`) but I saw that it introduces more allocations (or memory use) than the existing `strings.ToLower()` so I changed that back, but forgot to put back this value here.

Let me update that.

aaronlehmann commented on frontend/dockerfile/shell/lex_test.go:
```diff
@@ -44,8 +42,7 @@ func TestShellParserMandatoryEnvVars(t *testing.T) {
 	require.Equal(t, "", newWord)
```
This could be simplified to `require.Empty`

thaJeztah replied:
Ah! Yes, let me update that as well. I'm not a frequent user of testify, so didn't go look for those. My eye landed on the string matching for errors, which felt "incorrect", so I assumed testify had a function for that (it did), which also produced better output on failures, so I updated them 👍

thaJeztah commented:
@aaronlehmann thanks for reviewing; I updated some of your suggestions, and also updated the commit message for the "types" commit (to outline what I commented above)  Let me know what you think (feedback welcome as always 👍 )

thaJeztah commented:
@tonistiigi @crazy-max ptal

thaJeztah commented:
Rebased, after https://github.com/moby/buildkit/pull/3305 was merged


@AkihiroSuda @crazy-max @tonistiigi PTAL

jedevc commented on frontend/dockerfile/instructions/bflag.go:
```diff
@@ -180,27 +173,27 @@ func (bf *BFlags) Parse() error {
+			switch strings.ToLower(value) {
+			case "true", "":
 				flag.Value = "true"
-			} else if lower == "true" || lower == "false" {
-				flag.Value = lower
-			} else {
+			case "false":
+				flag.Value = "false"
```
Not sure why we've not used `strconv.ParseBool` here? We should probably use that instead, to be consistent with everywhere else where we do that.

thaJeztah replied:
Possibly because we wanted to define our own API, and `strconv.ParseBool` is quite relaxed ("t, T, f, F, 1, 0, true, false, TRUE, FALSE, True, False" are all accepted), which can become confusing quickly; do you know if making it more relaxed would affect caching? (e.g. changing `foo=true` to `foo=TRUE` now causing a cache-miss)?

jedevc replied:
It shouldn't affect caching, but agreed, it is an API change, and I'm not sure the alternative syntaxes are any easier to read/write - let's leave as is for now.

jedevc commented on frontend/dockerfile/instructions/commands_runmount.go:
```diff
@@ -171,18 +158,16 @@ func parseMount(value string, expander SingleWordExpander) (*Mount, error) {
 		}
 
-		if len(parts) != 2 {
+		if !ok {
```
We can merge this into the previous `if`.

thaJeztah commented on frontend/dockerfile/instructions/commands_runmount.go:
```diff
@@ -241,10 +233,11 @@ func parseMount(value string, expander SingleWordExpander) (*Mount, error) {
 		case "sharing":
-			if _, ok := allowedSharingTypes[strings.ToLower(value)]; !ok {
-				return nil, errors.Errorf("unsupported sharing value %q", value)
+			v := ShareMode(strings.ToLower(value))
+			if _, ok := allowedSharingModes[v]; !ok {
+				return nil, suggest.WrapError(errors.Errorf("unsupported sharing value %q", value), value, allShareModes(), true)
```
I based this on df9781b46c4ed4f4e7146c2241c0cb513167c2c5, but am slightly confused, as the `suggest.WrapError()` uses `caseSensitive=true`, but the code here suggests otherwise (perhaps this was meant to be an optimisation to not have to do the `strings.ToLower()`? But in that case we should pass `v` to the function, not `value`.

thaJeztah reviewed (commented):
@jedevc updated;

- changed types from an alias to a strong type
- merged the `!ok` check

I also pushed 2 additional commits;

- frontend/dockerfile: provide suggestions for mount share mode
- frontend/dockerfile: move check for cache-sharing

crazy-max reviewed (approved):
LGTM, ptal @jedevc @tonistiigi

jedevc commented on frontend/dockerfile/instructions/commands_runmount.go:
```diff
@@ -309,5 +305,9 @@ func parseMount(value string, expander SingleWordExpander) (*Mount, error) {
 	}
 
+	if m.CacheSharing != "" && m.Type != MountTypeCache {
+		return nil, errors.Errorf("invalid cache sharing set for %v mount", m.Type)
```
We can remove the above `m.Type == MountTypeSecret` variant of this check entirely, and replace this with:

```go
		return nil, errors.Errorf("%s mount should not define sharing set", m.Type)
```

or similar.

thaJeztah replied:
oh, I see I missed your comment; let me know if you want me to look at this
